File: use-cases/data-kiosk/code/java/src/main/java/lambda/CreateQueryHandler.java
package lambda;

import com.amazonaws.services.lambda.runtime.Context;
import com.amazonaws.services.lambda.runtime.LambdaLogger;
import com.amazonaws.services.lambda.runtime.RequestHandler;
import com.google.gson.Gson;
import io.swagger.client.api.QueriesApi;
import io.swagger.client.model.datakiosk.CreateQueryResponse;
import io.swagger.client.model.datakiosk.CreateQuerySpecification;
import lambda.utils.ApiUtils;

import java.util.Map;

import static lambda.utils.Constants.QUERY_CODE_KEY_NAME;
import static lambda.utils.Constants.REFRESH_TOKEN_ARN_ENV_VARIABLE;
import static lambda.utils.Constants.REGION_CODE_ARN_ENV_VARIABLE;

public class CreateQueryHandler implements RequestHandler<Map<String, String>, CreateQueryResponse> {

    // Sample event input:
    // {
    //  "Query": "query MyQuery{analytics_salesAndTraffic_2023_11_15{salesAndTrafficByAsin...",
    // }

    public CreateQueryResponse handleRequest(Map<String, String> input, Context context) {
        LambdaLogger logger = context.getLogger();
        logger.log("CreateQuery Lambda input: " + new Gson().toJson(input));

        // Retrieve request details from the input payload
        String queryCode = input.get(QUERY_CODE_KEY_NAME);
        String regionCode = System.getenv(REGION_CODE_ARN_ENV_VARIABLE);
        String refreshToken = System.getenv(REFRESH_TOKEN_ARN_ENV_VARIABLE);

        try {
            // Initialize the DataKiosk API client
            QueriesApi dataKioskApi = ApiUtils.getDataKioskApi(regionCode, refreshToken);

            // Build the createQuery request using the query from the input payload
            CreateQuerySpecification request = new CreateQuerySpecification();
            request.setQuery(queryCode);

            // Call the createQuery method to cancel the query
            CreateQueryResponse response = dataKioskApi.createQuery(request);
            logger.log(String.format("CreateQuery API response: %s", new Gson().toJson(response)));

            return response;
        } catch (Exception e) {
            throw new InternalError("Create query failed", e);
        }
    }

}

File: use-cases/data-kiosk/code/python/src/create_query_handler.py
import logging
import os

from utils.api_utils import ApiUtils
from utils import constants

logger = logging.getLogger()
logger.setLevel(logging.INFO)

"""
{
  "Query": "query MyQuery{analytics_salesAndTraffic_2023_11_15{salesAndTrafficByAsin...",
}
"""


def lambda_handler(event, context):
    # Retrieve request details from the input payload
    query_code = event[constants.QUERY_CODE_KEY_NAME]
    region_code = os.environ.get(constants.REGION_CODE_ARN_ENV_VARIABLE)
    refresh_token = os.environ.get(constants.REFRESH_TOKEN_ARN_ENV_VARIABLE)

    # Log the input event for the Lambda
    logger.info(f"CreateQuery Lambda input: {event}")

    try:
        # Create an instance of the ApiUtils class
        api_utils = ApiUtils(refresh_token, region_code, constants.DATA_KIOSK_API_TYPE)

        # Build the createQuery request using the query from the input payload
        body = {
            "query": query_code
        }

        # Call the create_query endpoint to create the query
        create_query_result = api_utils.call_datakiosk_api('create_query', body=body)

        logger.info(f"CreateQuery API Response: {create_query_result}")

        return create_query_result.to_dict()

    except Exception as e:
        raise Exception("Calling Data Kiosk API failed", e.__dict__)
